### DISCUSSION TITLE
Unable to override cookie policy

### DISCUSSION THREAD

**Discussion by KKomarov** [encode/httpx#2950] [General]
I would like to be able to override the cookie policy used by a Client when making a request. However, overriding the cookie policy on the cookie jar of the Client or of the Request object has had no effect. These lines of code explains why:

https://github.com/encode/httpx/blob/497b315fc7a0ca6a54c8ec6c929568abaf6efab1/httpx/_models.py#L334
https://github.com/encode/httpx/blob/497b315fc7a0ca6a54c8ec6c929568abaf6efab1/httpx/_models.py#L1030
When initialise request from existing cookies (which is Cookies object of Client), it discards policy object from source Cookies object.

https://github.com/encode/httpx/blob/497b315fc7a0ca6a54c8ec6c929568abaf6efab1/httpx/_client.py#L400
The cookie jars of both the client and the request are merged into a newly constructed CookieJar, discarding the policies of both. 

What I expected to be able to do, and what isn't working currently, is something like the following:
```
        cookie = MozillaCookieJar(filename='cookies.txt', policy=MyCookiePolicy())
        cookie.load()
        session = httpx.Client(headers=headers, cookies=cookie)
        session.get(url)
```
In my opinion Policy from CookieJar in Client should not be discarded. When CookieJar object passed in request then policy from request should override policy from Client.

**Comment by lovelydinosaur** [comment]
Interesting question. There's not much out there on Python's cookie policy API, so I'd like to start by understanding your use case more thoroughly first... what is the alternate policy that you'd actually like to set, and how does it differ from the default?

**Reply by KKomarov** [reply]
Cookie policy is just a filter on what cookies from cookiejar you want to add in actual request. I was trying to modify default policy to include cookies from www subdomain to main domain request. Also I want to add session cookies to request, default cookie policy treats them as expired and discards. It was quite unexpected behaviour to me accept CookieJar but using only half. For reference we might have a look at requests lib, they having similar behaviour. There was several discussions in their repo and they fixed this in 3.0.0 (which has never released actually)

**Reply by hyosotis** [reply]
In my use case, I want to share cookies between domains.

I do it by this way when i use aiohttp.
```python
class CookieJar(aiohttp.abc.AbstractCookieJar):

    def update_cookies(self, cookies: aiohttp.typedefs.LooseCookies, response_url: yarl.URL = ...):
        ...

s = aiohttp.ClientSession(cookie_jar=CookieJar())
```

but in httpx.
I can't provide custom CookieJar.
https://github.com/encode/httpx/blob/5b5f6d8e172910371b5901f6e21717097274c513/httpx/_client.py#L183
